pinot-core/src/main/java/org/apache/pinot/core/data/table/OrderedIndexedTableResizer.java

```diff
@@ -0,0 +1,260 @@
+      int size;
+      Comparator<IntermediateRecord> comparator;
+
+      if (numRecordsToEvict < trimToSize) { // num records to evict is smaller than num records to retain
+        size = numRecordsToEvict;
+        // make PQ of records to evict
+        comparator = _intermediateRecordComparator;
+      } else { // num records to retain is smaller than num records to evict
+        size = trimToSize;
+        // make PQ of records to retain
+        comparator = _intermediateRecordComparator.reversed();
+      }
+
+      PriorityQueue<IntermediateRecord> priorityQueue =
+          new PriorityQueue<>(size, comparator);
+
+      for (Record record : recordsMap.values()) {
+
+        IntermediateRecord intermediateRecord = getIntermediateRecord(record);
+        if (priorityQueue.size() < size) {
+          priorityQueue.offer(intermediateRecord);
+        } else {
+          IntermediateRecord peek = priorityQueue.peek();
+          if (priorityQueue.comparator().compare(peek, intermediateRecord) < 0) {
+            priorityQueue.poll();
+            priorityQueue.offer(intermediateRecord);
+          }
+        }
+      }
+
+      if (numRecordsToEvict < trimToSize) { // PQ contains records to evict
+        for (IntermediateRecord evictRecord : priorityQueue) {
+          recordsMap.remove(evictRecord.getKey());
+        }
+      } else { // PQ contains records to retain
+        Set<Key> keysToRetain = new HashSet<>(priorityQueue.size());
+        for (IntermediateRecord retainRecord : priorityQueue) {
+          keysToRetain.add(retainRecord.getKey());
+        }
+        recordsMap.keySet().retainAll(keysToRetain);
+      }
+    }
+  }
+
+  /**
+   * Sort the records in the recordsMap according to order by
+   */
+  @Override
+  List<Record> sortRecordsMap(Map<Key, Record> recordsMap) {
+    int numRecords = recordsMap.size();
+    if (numRecords == 0) {
+      return Collections.emptyList();
+    }
+    List<Record> sortedRecords = new ArrayList<>(numRecords);
+    List<IntermediateRecord> intermediateRecords = new ArrayList<>(numRecords);
+    for (Record record : recordsMap.values()) {
+      intermediateRecords.add(getIntermediateRecord(record));
+    }
+    intermediateRecords.sort(_intermediateRecordComparator);
+    for (IntermediateRecord intermediateRecord : intermediateRecords) {
+      sortedRecords.add(recordsMap.get(intermediateRecord.getKey()));
+    }
+    return sortedRecords;
+  }
+
+  /**
+   * Enum to differentiate between different types of order by
+   * Order by can be on the key column (group by values or distinct), or on the aggregation column
+   */
+  private enum OrderByType {
+    KEY_COLUMN, // order by on a key column (e.g. group by or distinct)
+    AGGREGATION_COLUMN // order by on aggregation (e.g. aggregations in aggregation only or group by)
+  }
+
+  /**
+   * Defines the information needed for one SelectionSort of an order by clause
+   */
+  private class OrderByInfo {
```

Comment:
This class can be replaced with the OrderByValueExtractor to avoid condition checks